consensus/abci.go

```diff
@@ -123,17 +122,50 @@ func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx
 	
 	tx, err := decodeRLP(txBytes)
 	if err != nil {
-		abci.logger.Error("Received invalid transaction", "tx", tx.Hash().String())
-		abci.metrics.CheckErrTxsTotal.Add(1, fmt.Sprint(abciTypes.ErrEncodingError.Code))
-		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrEncodingError.Code), Log: err.Error()}
+		abci.logger.Error("Unable to decode RLP TX", "err", err.Error())
+		abci.metrics.CheckErrTxsTotal.Add(1, "INVALID_TX")
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: "INVALID_TX"}
 	}
 
 	abci.logger.Info("Checking TX", "hash", tx.Hash().String(), "nonce", tx.Nonce(), "cost", tx.Cost())
+	from, err := abci.doMempoolValidation(tx)
+	if err != nil {
+		abci.logger.Error(err.Error())
+		abci.metrics.CheckErrTxsTotal.Add(1, err.Error())
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: err.Error()}
+	}
+
+	// Additional custom validation aside of the Ethereum default mempool validation
+	if tx.GasPrice().Uint64() < abci.db.Config().TxPool.PriceLimit {
+		abci.logger.Error("TX gas price too low", "tx_gas_price", tx.GasPrice(), "required_tx_gas_price", abci.db.Config().TxPool.PriceLimit)
+		abci.metrics.CheckErrTxsTotal.Add(1, "INSUFFICIENT_GAS_PRICE")
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: "INSUFFICIENT_GAS_PRICE"}
+	}
+
+	currentBalance := abci.checkTxState.GetBalance(from)
+	txCost := tx.Cost()
+	if currentBalance.Cmp(txCost) < 0 {
+		abci.logger.Error("Low balance", "balance", currentBalance.String(), "tx_cost", tx.Cost().String())
+		abci.metrics.CheckErrTxsTotal.Add(1, "INSUFFICIENT_FUNDS")
+		return tmtAbciTypes.ResponseCheckTx{Code: 1, Log: "INSUFFICIENT_FUNDS"}
+	}
+
+	// Adjust From and To balances in order for the balance verification to be valid for next TX
+	abci.checkTxState.SubBalance(from, txCost)
+	if to := tx.To(); to != nil {
+		abci.checkTxState.AddBalance(*to, tx.Value())
+	}
+
+	abci.checkTxState.SetNonce(from, tx.Nonce() + 1)
+
+	abci.logger.Info("TX validated.", "hash", tx.Hash().String(), "state_nonce", abci.checkTxState.GetNonce(from))
+
+	return tmtAbciTypes.ResponseCheckTx{Code: tmtAbciTypes.CodeTypeOK}
+}
 
+func (abci *TendermintABCI) doMempoolValidation(tx *ethTypes.Transaction) (address common.Address, err error) {
```

Comment:
[MAJOR] I cannot understand why a method responsible for doing a validation returns a relevant value such as `from` account back. You clearly break the single responsibility propose of this method.

Reply:
ahhh ok